staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,330 +17,238 @@ limitations under the License.
+		if result.Err != nil {
+			return result
+		}
+		return cached.NewResultOK(aggregator.FilterSpecByPathsWithoutSideEffects(result.Data, []string{"/apis/"}), result.Etag)
+	}, result)
+	specInfo.spec.Replace(filteredResult)
+	return result.Err
 }
 
-// UpdateAPIServiceSpec updates the api service's OpenAPI spec. It is thread safe.
-func (s *specAggregator) UpdateAPIServiceSpec(apiServiceName string, spec *spec.Swagger, etag string) error {
-	s.rwMutex.Lock()
-	defer s.rwMutex.Unlock()
-
-	specInfo, existingService := s.openAPISpecs[apiServiceName]
-	if !existingService {
-		return fmt.Errorf("APIService %q does not exists", apiServiceName)
-	}
-
-	// For APIServices (non-local) specs, only merge their /apis/ prefixed endpoint as it is the only paths
-	// proxy handler delegates.
-	if specInfo.apiService.Spec.Service != nil {
-		spec = aggregator.FilterSpecByPathsWithoutSideEffects(spec, []string{"/apis/"})
-	}
-
-	return s.tryUpdatingServiceSpecs(&openAPISpecInfo{
-		apiService: specInfo.apiService,
-		spec:       spec,
-		handler:    specInfo.handler,
-		etag:       etag,
-	})
+// UpdateAPIService updates the api services.. It is thread safe.
+func (s *specAggregator) UpdateAPIService(apiServiceName string) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return s.updateServiceLocked(apiServiceName)
 }
 
-// AddUpdateAPIService adds or updates the api service. It is thread safe.
-func (s *specAggregator) AddUpdateAPIService(handler http.Handler, apiService *v1.APIService) error {
-	s.rwMutex.Lock()
-	defer s.rwMutex.Unlock()
-
+// AddAPIService adds the api service. It is thread safe. If the
+// apiservice already exists, it will be updated.
+func (s *specAggregator) AddAPIService(apiService *v1.APIService, handler http.Handler) error {
 	if apiService.Spec.Service == nil {
-		// All local specs should be already aggregated using local delegate chain
 		return nil
 	}
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
 
-	newSpec := &openAPISpecInfo{
-		apiService: *apiService,
-		handler:    handler,
-	}
-	if specInfo, existingService := s.openAPISpecs[apiService.Name]; existingService {
-		newSpec.etag = specInfo.etag
-		newSpec.spec = specInfo.spec
-	}
-	return s.tryUpdatingServiceSpecs(newSpec)
-}
-
-// RemoveAPIServiceSpec removes an api service from OpenAPI aggregation. If it does not exist, no error is returned.
-// It is thread safe.
-func (s *specAggregator) RemoveAPIServiceSpec(apiServiceName string) error {
-	s.rwMutex.Lock()
-	defer s.rwMutex.Unlock()
-
-	if _, existingService := s.openAPISpecs[apiServiceName]; !existingService {
-		return nil
+	_, exists := s.openAPISpecs[apiService.Name]
+	if !exists {
+		s.openAPISpecs[apiService.Name] = &openAPISpecInfo{
+			apiService: *apiService,
+			downloader: newDecoratedCache(apiService.Name, NewCacheableDownloader(s.downloader, handler)),
+		}
+		// Re-create the mergeSpec for the new list of apiservices
```

Comment:
"Recreate" is a big word for promise-style `cached.Data[X]` return types :)

Reply:
Relink? Reconnect? Update the cache graph? Not sure how to best word this @apelisse

Reply:
the code is self-described IMO, I think you can drop the comment, i.e.
```golang
s.openAPIVersionedService.UpdateSpecLazy(s.buildMergeSpecLocked())
```
says it all.